azplugins/BondEvaluatorDoubleWell.h

```diff
@@ -0,0 +1,135 @@
+#endif
+
+namespace azplugins
+{
+namespace detail
+{
+
+//! Class for evaluating the double well bond potential
+/*!
+This bond potential follows the functional form
+\f{eqnarray*}
+V_{\mathrm{DW}}(r)  =   \frac{V_{min}}{b^4} \left[ \left( r - a/2 \right)^2 -b^2 \right]^2
+\f}
+and has two minima at r = 1/2(a +/- 2b), seperated by a maximum at 1/2a of height V_max.
+The parameter a tunes the location of the maximal value and the parameter b tunes the distance of the
+two maxima from each other.  This potential is useful to model bonds which can be either mechanically or
+thermally "activated" into a effectively longer state. The value of V_max can be used to tune the height of the
+energy barrier in between the two states.
+
+The parameters are:
+    - \a V_max (params.x) maximum potential value between the two minima
+    - \a a (params.y) shift for the location of the V_max, maximun is at 1/2 a
+    - \a b (params.z) scaling for the distance of the two minima at 1/2(a +/- 2b)
+*/
+class BondEvaluatorDoubleWell
+    {
+    public:
+        //! Define the parameter type used by this bond potential evaluator
+        typedef Scalar3 param_type;
+
+        //! Constructs the pair potential evaluator
+        /*!
+         * \param _rsq Squared distance beteen the particles
+         * \param _params Per type bond parameters of this potential as given above
+         */
+        DEVICE BondEvaluatorDoubleWell(Scalar _rsq, const param_type& _params)
+            : rsq(_rsq), V_max(_params.x), a(_params.y), b(_params.z)
+            {
+            }
+
+        //! This evaluator doesn't use diameter information
+        DEVICE static bool needsDiameter() { return false; }
+
+        //! Accept the optional diameter values
+        /*!
+         * \param da Diameter of particle a
+         * \param db Diameter of particle b
+         */
+        DEVICE void setDiameter(Scalar da, Scalar db) {  }
+
+        //! FENE doesn't use charge
+        DEVICE static bool needsCharge() { return false; }
+
+        //! Accept the optional charge values
+        /*!
+         * \param qa Charge of particle a
+         * \param qb Charge of particle b
+         */
+        DEVICE void setCharge(Scalar qa, Scalar qb) { }
+
+        //! Evaluate the force and energy
+        /*!
+         * \param force_divr Output parameter to write the computed force divided by r.
+         * \param bond_eng Output parameter to write the computed bond energy
+         *
+         *  \return True if they are evaluated or false if the bond energy is not defined.
+         */
+        DEVICE bool evalForceAndEnergy(Scalar& force_divr, Scalar& bond_eng)
+            {
+            // check for invalid parameters
+            if (b == Scalar(0.0)) return false;
+
+            Scalar r = sqrt(rsq);
+            Scalar r_min_half_a = r-0.5*a;
+            Scalar b_sq = b*b;
+
+            bond_eng = V_max/(b_sq*b_sq)*(r_min_half_a*r_min_half_a - b_sq)*(r_min_half_a*r_min_half_a - b_sq);
```

Comment:
I would recommend saving `r_min_half_a*rmin_half_a - b_sq` in a const variable since you recycle it three times, so it will make it clear to the compiler that the result should be reused.